compiler/rustc_resolve/src/build_reduced_graph.rs

```diff
@@ -244,17 +265,31 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
         child: &ModChild,
         parent_scope: ParentScope<'ra>,
         child_index: usize,
+        ambig_child: Option<(&ModChild, AmbigModChildKind)>,
     ) {
         let parent = parent_scope.module;
+        let child_span = |this: &Self, reexport_chain: &[Reexport], res: def::Res<_>| {
+            this.def_span(
+                reexport_chain
+                    .first()
+                    .and_then(|reexport| reexport.id())
+                    .unwrap_or_else(|| res.def_id()),
+            )
+        };
         let ModChild { ident, res, vis, ref reexport_chain } = *child;
-        let span = self.def_span(
-            reexport_chain
-                .first()
-                .and_then(|reexport| reexport.id())
-                .unwrap_or_else(|| res.def_id()),
-        );
+        let span = child_span(self, reexport_chain, res);
         let res = res.expect_non_local();
         let expansion = parent_scope.expansion;
+        let ambig = ambig_child.map(|(ambig_child, ambig_kind)| {
+            let ModChild { ident: _, res, vis, ref reexport_chain } = *ambig_child;
+            let span = child_span(self, reexport_chain, res);
+            let res = res.expect_non_local();
+            let ambig_kind = match ambig_kind {
+                AmbigModChildKind::GlobVsGlob => AmbiguityKind::GlobVsGlob,
+                AmbigModChildKind::GlobVsExpanded => AmbiguityKind::GlobVsExpanded,
+            };
+            (self.arenas.new_res_binding(res, vis, span, expansion), ambig_kind)
+        });
         // Record primary definitions.
         match res {
```

Comment:
NIT: When I was reviewing this section I felt like there was a high signal to noise ratio with the way this match is structured. As far as I can tell the only difference between the different `define_extern` calls is the namespace argument. I'd recommend changing this match to evaluate to the namespace and pass that ns local into a single shared invocation of `define_extern` regardless of res defkind.

Especially with the new field changing the formatting to be one line per argument, its even more difficult now to visually tell what is different and what isn't between those invocations.

Feel free to ignore this one, just thought I'd share my experience.

Reply:
I can factor the `define_extern` calls into a closure.
I'd like to preserve the exhaustive match because of the `bug!("unexpected resolution: {:?}", res)` part.